ratelimit/redis.go

```diff
@@ -0,0 +1,82 @@
+package ratelimit
+
+import (
+	"context"
+	_ "embed"
+	"strconv"
+	"time"
+
+	"github.com/redis/rueidis"
+)
+
+//go:embed gcra_rate_limit.lua
+var gcraRateLimitScript string
+
+// RedisRateLimiter is a type that represents a rate limiter that uses Redis as its backend.
+// The rate limiting is a leaky bucket implementation using the generic cell rate algorithm.
+// See https://en.wikipedia.org/wiki/Generic_cell_rate_algorithm for details on how this algorithm works.
+type RedisRateLimiter struct {
+	client rueidis.Client
+}
+
+// Ensure RedisRateLimiter implements the SharedRateLimiter interface.
+var _ SharedRateLimiter = (*RedisRateLimiter)(nil)
+
+// NewRedisRateLimiter creates a new instance of RedisRateLimiter.
+func NewRedisRateLimiter(addresses []string) (*RedisRateLimiter, error) {
+	client, err := rueidis.NewClient(rueidis.ClientOption{InitAddress: addresses})
+	if err != nil {
+		return nil, err
+	}
+	return &RedisRateLimiter{client: client}, nil
+}
+
+// GetRateLimits retrieves the rate limits for a given request using a Redis Rate Limiter.
+// It returns the amount of remaining requests, the reset time in milliseconds, and any error that occurred.
+func (r *RedisRateLimiter) GetRateLimits(ctx context.Context, req *request) (remaining, resetTime int64, err error) {
+	inspectScript := rueidis.NewLuaScript(gcraRateLimitScript)
+	rateLimitParameters := []string{
+		strconv.FormatInt(time.Now().UnixMilli(), 10), // now
+		strconv.FormatInt(req.limit, 10),              // burst
```

Comment:
Is this intended, we want the burst to equal the limit always? Or should it be 0?

Reply:
For now, this is intended. I didn't want to expose too much configuration yet. Later this can be made configurable too. 👍

Reply:
I misunderstood how it works. Clarified in latest PR with comments. Nothing to change